Improve deprecation for aud claims (#630)
lcobucci/jwt: docs/upgrading.md

To help on the migration process, all deprecated components are being marked with `@deprecated` and deprecated behaviour will trigger a `E_USER_DEPRECATED` error.
However, you can also find here the instructions on how to make your code compatible with both versions.

### General migration strategy

> It is recommended to update your existing software from any previous 3.x version to the latest 3.4.x version using 

Reviewer: Isn't this implied on the mention of the compatibility layer above? Perhaps we don't need a sub-section for the "General migration strategy" but add a paragraph above encouraging people to go through v3.4?

The remarks about PHPUnit are important but maybe using a note/warning might get more attention?

composer:

>     composer require lcobucci/jwt ^3.4

Reviewer: Please use the triple backslashes with `sh` as syntax highlight here.

Then run your tests and fix all deprecation notices.

Note that PHPUnit will only break tests if you have the `E_USER_DEPRECATED` error level activated (it is a good idea to
run tests using `E_ALL`, btw.), and deprecation messages during the execution of data providers will not affect the 
test result at all, only output text to the console, because they are called before test execution starts. Make sure 
you do not see any of them before you continue.

Now you can upgrade to the latest 4.x version:

    composer require lcobucci/jwt ^4.0

At this stage, you'll probably find some more rough edges. Not every method call in the 3.4 version is
100% identical to the 4.0 counterpart. So feel free to optimize your code at this point to leverage the
core improvements like the configuration object, validation being separated from the `Token` value object and such. 

> The details are listed below.

Reviewer: Is there any rough edge that you've found apart from the cases documented in this file? If not, why this paragraph is relevant?

The idea of this doc section is to be a step-by-step kind of thing, so people would go through the things below and modify their code. I'm confused with the need for summarising the changes.

Author: Rephrased and put emphasis on deprecation messages, and using the new features detailed later in this document.

> 

Reviewer: We (try) to use [one sentence per line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line) on the docs to make things a bit easier to maintain (especially regarding diffs in the future).

Author: Will reformat.

To help on the migration process, all deprecated components are being marked with `@deprecated` and deprecated behaviour will trigger a `E_USER_DEPRECATED` error.
However, you can also find here the instructions on how to make your code compatible with both versions.

### General migration strategy

Update your existing software to the latest 3.4.x version using composer:

```sh

>     composer require lcobucci/jwt ^3.4

Reviewer: ```suggestion
composer require lcobucci/jwt ^3.4
```

```

> Then run your tests and fix all deprecation notices.

Reviewer: ```suggestion
Then run your tests and tools like [`phpstan/phpstan-deprecation-rules`](https://github.com/phpstan/phpstan-deprecation-rules) to fix all deprecation notices.
```

Author: I'm struggling a bit with this part of the text because your suggestion is different from my experience, but I will add the tool link.

> Note that PHPUnit tests will only fail if you have the `E_USER_DEPRECATED` error level activated (it is a good idea to run tests using `E_ALL`, btw.), and deprecation messages during the execution of data providers will not affect the test result at all, only output text to the console, because they are called before test execution starts.

Reviewer: This sentence is a bit long... what do you think about this:

```suggestion
Note that PHPUnit tests will only fail if you have the `E_USER_DEPRECATED` error level activated - it is a good practice to run tests using `E_ALL`.
Deprecation messages during the execution of data providers will not affect the test result at all, only output text to the console, because they are called before test execution starts.
```

Make sure you do not see any of them before you continue.

Now you can upgrade to the latest 4.x version:

```sh

>     composer require lcobucci/jwt ^4.0

Reviewer: ```suggestion
composer require lcobucci/jwt ^4.0
```

```

Remember that some deprecation messages from the 3.4 version may have notified you that things still are different in 4.0, so you may find you need to adapt your own code once more at this stage.

> Not every method call in the 3.4 version is 100% identical to the 4.0 counterpart.

Reviewer: @SvenRtbg wasn't that mitigated with the backport of multiple audiences? Do you have other examples that we need to address?

I do expect that code that uses the forward compatibility layer should simply work on v4.x.

Author: I think I was referring to the fact that deprecated methods, or more precisely `getClaim('aud')`, changed the return value compared to 3.3. And that difference does still exist. For a deprecated method that should not be used... it's probably unwarranted to talk about an undetermined amount of methods that "may" be different.

Author: I opted for removing that sentence. I do not remember enough of my migration process to have proof for this statement, and I don't want to redo that work again. :)

Author: Is there a reason the 4.0 thru 4.2 `Token` interface has no `claims()` method defined? I'm comparing 3.4 `Token` and 4.0 `Token` side-by-side right now to see if the method signatures match (reasonably).

Author: `payload()` and `signature()` also seem to be missing.

Author: I found them in the 4.2 variant...

#### Support for multiple audiences

Even though we didn't officially support multiple audiences, it was technically possible to achieve that by manually setting the `aud` claim to an array with multiple strings.

If you parse a token with 3.4, and read it's contents with `\Lcobucci\JWT\Token#getClaim()` or`\Lcobucci\JWT\Token#getClaims()`, you will only get the first element of such an array back.
If the audience claim does only contain a string, or only contains one string in the array, nothing changes.
Please [upgrade to the new Token API](#use-the-new-token-api) for accessing claims in order to get the full audience array again (e.g. call `Token#claims()->get('aud')`).

> When creating a token, use the new method `Builder#permittedFor()` as detailed below:

Reviewer: ```suggestion
When creating a token, use the new method `Builder#permittedFor()` as detailed below.
```

influence this by adding or removing the `\Lcobucci\JWT\Encoding\UnifyAudience` claims formatter class, or your own 
implementation.

In 3.4 however there is no such thing available, and the implementation hardcoded casts the argument passed to
`\Lcobucci\JWT\Builder::permittedFor()` as a string (forcing an array to become the string "Array" and triggering the
"array to string conversion" PHP notice), wrapping it inside an array, with no way to add more than one string to the 
token. Reading from the token via `\Lcobucci\JWT\Token::getClaim()` or`\Lcobucci\JWT\Token::getClaims()` will revert 
this array wrapping and only ever return the first entry in the "aud" claims array, if it is one, or return the string 
otherwise.

If you pass an array to the builder or if you read an array from the token, a deprecation message will be
triggered to inform you of this behaviour change. There is no workaround for version 3.4 if you have to
access or create "aud" claims that are arrays with multiple entries.

In 4.0, the `permittedFor()` method accepts a variadic parameter, so you can either call 

> `permittedFor("audience1", "audience2")` or `permittedFor(...["audience1", "audience2"])` 

Reviewer: This assumes that the library had official support for multiple audiences on v3, which is not the case.

I'm happy to improve the docs to mention that things accidentally used to work because of lack of type declaration and loose validation on custom claim configuration, though.

Author: I think I found the hidden reason, and illustrated it in the new version of this part.
